File: backend/app/api/v1/endpoints/colaboradores.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from app.db.session import get_db
from app.core.deps import require_logistica, get_current_colaborador
from app.models.models import Coordenador, CertameColaborador, Usuario
from app.api.v1.schemas.colaboradores import (
    ColaboradorPreCadastro, ColaboradorCompletarCadastro, ColaboradorLogin,
    ColaboradorTokenResponse, ConviteInfo, ColaboradorAdminResponse,
    CertameSimples, VincularCertame, ColaboradorPortalPerfil,
)
from app.api.v1.services.colaboradores import (
    pre_cadastrar_colaborador, verificar_convite, completar_cadastro,
    login_colaborador, listar_colaboradores, obter_colaborador,
    deletar_colaborador, regenerar_convite, vincular_certame,
    desvincular_certame, listar_certames_colaborador,
)

router = APIRouter()


def _to_admin_response(colab: Coordenador) -> ColaboradorAdminResponse:
    certames = [
        CertameSimples(
            id=v.id,
            certame_id=v.certame_id,
            titulo=v.certame.titulo,
            orgao=v.certame.orgao,
            funcao=v.funcao,
        )
        for v in colab.certames_vinculados
    ]
    return ColaboradorAdminResponse(
        id=colab.id,
        nome=colab.nome,
        cpf=colab.cpf,
        celular=colab.celular,
        email=colab.email,
        status=colab.status.value,
        token_convite=colab.token_convite,
        token_expiry=colab.token_expiry,
        criado_em=colab.criado_em,
        certames=certames,
    )


# ── Rotas públicas ────────────────────────────────────────────────────────────

@router.get("/colaboradores/convite/{token}", response_model=ConviteInfo, tags=["portal-colaborador"])
def get_convite(token: str, db: Session = Depends(get_db)):
    return verificar_convite(db, token)


@router.post("/colaboradores/completar-cadastro/{token}", response_model=ConviteInfo, tags=["portal-colaborador"])
def post_completar_cadastro(
    token: str,
    data: ColaboradorCompletarCadastro,
    db: Session = Depends(get_db),
):
    return completar_cadastro(db, token, data)


@router.post("/colaboradores/login", response_model=ColaboradorTokenResponse, tags=["portal-colaborador"])
def post_login_colaborador(data: ColaboradorLogin, db: Session = Depends(get_db)):
    return login_colaborador(db, data)


# ── Rotas portal (token de colaborador) ──────────────────────────────────────

@router.get("/colaboradores/portal/perfil", response_model=ColaboradorPortalPerfil, tags=["portal-colaborador"])
def get_portal_perfil(current: Coordenador = Depends(get_current_colaborador)):
    return current


@router.get("/colaboradores/portal/certames", tags=["portal-colaborador"])
def get_portal_certames(
    db: Session = Depends(get_db),
    current: Coordenador = Depends(get_current_colaborador),
):
    vinculos = listar_certames_colaborador(db, current.id)
    return [
        {
            "id": v.id,
            "certame_id": v.certame_id,
            "titulo": v.certame.titulo,
            "orgao": v.certame.orgao,
            "data_aplicacao": v.certame.data_aplicacao,
            "status": v.certame.status.value,
            "funcao": v.funcao,
        }
        for v in vinculos
    ]


# ── Rotas admin ───────────────────────────────────────────────────────────────

@router.get("/colaboradores", response_model=list[ColaboradorAdminResponse], tags=["colaboradores"])
def get_colaboradores(
    db: Session = Depends(get_db),
    _: Usuario = Depends(require_logistica),
):
    return [_to_admin_response(c) for c in listar_colaboradores(db)]


@router.post("/colaboradores/pre-cadastro", response_model=ColaboradorAdminResponse, status_code=201, tags=["colaboradores"])
def post_pre_cadastro(
    data: ColaboradorPreCadastro,
    db: Session = Depends(get_db),
    _: Usuario = Depends(require_logistica),
):
    colab = pre_cadastrar_colaborador(db, data)
    return _to_admin_response(colab)


@router.delete("/colaboradores/{colab_id}", status_code=204, tags=["colaboradores"])
def delete_colaborador(
    colab_id: str,
    db: Session = Depends(get_db),
    _: Usuario = Depends(require_logistica),
):
    deletar_colaborador(db, colab_id)


@router.post("/colaboradores/{colab_id}/reenviar-convite", response_model=ColaboradorAdminResponse, tags=["colaboradores"])
def post_reenviar_convite(
    colab_id: str,
    db: Session = Depends(get_db),
    _: Usuario = Depends(require_logistica),
):
    colab = regenerar_convite(db, colab_id)
    return _to_admin_response(colab)


@router.post("/colaboradores/{colab_id}/vincular-certame", status_code=201, tags=["colaboradores"])
def post_vincular_certame(
    colab_id: str,
    data: VincularCertame,
    db: Session = Depends(get_db),
    _: Usuario = Depends(require_logistica),
):
    v = vincular_certame(db, colab_id, data)
    return CertameSimples(
        id=v.id,
        certame_id=v.certame_id,
        titulo=v.certame.titulo,
        orgao=v.certame.orgao,
        funcao=v.funcao,
    )


@router.get("/certames/{certame_id}/equipe", tags=["colaboradores"])
def get_equipe_certame(
    certame_id: str,
    db: Session = Depends(get_db),
    _: Usuario = Depends(require_logistica),
):
    vinculos = db.query(CertameColaborador).filter(
        CertameColaborador.certame_id == certame_id
    ).all()
    return [
        {
            "id": v.id,
            "certame_id": v.certame_id,
            "colaborador_id": v.coordenador_id,
            "funcao": v.funcao,
            "nome": v.colaborador.nome,
            "cpf": v.colaborador.cpf,
            "celular": v.colaborador.celular,
            "email": v.colaborador.email,
            "status": v.colaborador.status.value,
        }
        for v in vinculos
    ]


@router.delete("/colaboradores/{colab_id}/certames/{certame_id}", status_code=204, tags=["colaboradores"])
def delete_vinculo_certame(
    colab_id: str,
    certame_id: str,
    db: Session = Depends(get_db),
    _: Usuario = Depends(require_logistica),
):
    desvincular_certame(db, colab_id, certame_id)

File: src/services/colaboradores.ts
import api from './api'
import axios from 'axios'

// API pública (sem autenticação)
const publicApi = axios.create({ baseURL: '/api/v1' })

// API do portal do colaborador
const portalApi = axios.create({ baseURL: '/api/v1' })
portalApi.interceptors.request.use((config) => {
  const token = localStorage.getItem('glc_portal_token')
  if (token) config.headers.Authorization = `Bearer ${token}`
  return config
})
portalApi.interceptors.response.use(
  (r) => r,
  (error) => {
    if (error.response?.status === 401) {
      localStorage.removeItem('glc_portal_token')
      localStorage.removeItem('glc_portal_id')
      localStorage.removeItem('glc_portal_nome')
      window.location.href = '/portal/colaborador/login'
    }
    return Promise.reject(error)
  }
)

export interface ColaboradorAdmin {
  id: string
  nome: string
  cpf?: string
  celular?: string
  email?: string
  status: 'pendente' | 'ativo' | 'inativo'
  token_convite?: string
  token_expiry?: string
  criado_em: string
  certames: ColaboradorCertame[]
}

export interface ColaboradorCertame {
  id: string
  certame_id: string
  titulo: string
  orgao?: string
  funcao?: string
}

export interface ConviteInfo {
  nome: string
  cpf: string
  status: string
}

export interface CertamePortal {
  id: string
  certame_id: string
  titulo: string
  orgao?: string
  data_aplicacao?: string
  status: string
  funcao?: string
}

export interface MembroEquipe {
  id: string
  certame_id: string
  colaborador_id: string
  funcao?: string | null
  nome: string
  cpf?: string | null
  celular?: string | null
  email?: string | null
  status: 'pendente' | 'ativo' | 'inativo'
}

// ── Admin ────────────────────────────────────────────────────────────────────

export const colaboradoresAdminService = {
  listar: async (): Promise<ColaboradorAdmin[]> => {
    const { data } = await api.get('/colaboradores')
    return data
  },

  preCadastrar: async (payload: { nome: string; cpf: string; celular: string }): Promise<ColaboradorAdmin> => {
    const { data } = await api.post('/colaboradores/pre-cadastro', payload)
    return data
  },

  deletar: async (id: string): Promise<void> => {
    await api.delete(`/colaboradores/${id}`)
  },

  reenviarConvite: async (id: string): Promise<ColaboradorAdmin> => {
    const { data } = await api.post(`/colaboradores/${id}/reenviar-convite`)
    return data
  },

  vincularCertame: async (id: string, certame_id: string, funcao?: string): Promise<ColaboradorCertame> => {
    const { data } = await api.post(`/colaboradores/${id}/vincular-certame`, { certame_id, funcao })
    return data
  },

  desvincularCertame: async (id: string, certame_id: string): Promise<void> => {
    await api.delete(`/colaboradores/${id}/certames/${certame_id}`)
  },
}

// ── Equipe por certame ────────────────────────────────────────────────────────

export const certameEquipeService = {
  listar: async (certameId: string): Promise<MembroEquipe[]> => {
    const { data } = await api.get(`/certames/${certameId}/equipe`)
    return data
  },

  vincular: async (colabId: string, certameId: string, funcao?: string): Promise<ColaboradorCertame> => {
    return colaboradoresAdminService.vincularCertame(colabId, certameId, funcao)
  },

  desvincular: async (vinculoColabId: string, certameId: string): Promise<void> => {
    return colaboradoresAdminService.desvincularCertame(vinculoColabId, certameId)
  },
}

// ── Convite / Portal ──────────────────────────────────────────────────────────

export const conviteService = {
  verificar: async (token: string): Promise<ConviteInfo> => {
    const { data } = await publicApi.get(`/colaboradores/convite/${token}`)
    return data
  },

  completar: async (token: string, payload: object): Promise<ConviteInfo> => {
    const { data } = await publicApi.post(`/colaboradores/completar-cadastro/${token}`, payload)
    return data
  },

  login: async (cpf: string, senha: string): Promise<{ access_token: string; colaborador_id: string; nome: string }> => {
    const { data } = await publicApi.post('/colaboradores/login', { cpf, senha })
    return data
  },
}

export const portalService = {
  perfil: async (): Promise<ConviteInfo & { id: string; celular?: string; email?: string }> => {
    const { data } = await portalApi.get('/colaboradores/portal/perfil')
    return data
  },

  certames: async (): Promise<CertamePortal[]> => {
    const { data } = await portalApi.get('/colaboradores/portal/certames')
    return data
  },
}
